validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,265 @@
-                    .map(|result_all| {
-                        result_all
-                            .and_then(|all| Ok(all.into_iter().flatten().collect::<Vec<_>>()))
-                    })
-                    .boxed()
-            }
-        })
-        .boxed()
-}
-
-async fn fetch_page(
-    self,
-    page: u64,
-    validator: Option<ValidatorId>,
-) -> Result<ChannelAllResponse, reqwest::Error> {
-    let mut query = vec![format!("page={}", page)];
-
-    if let Some(validator) = validator {
-        query.push(format!("validator={}", validator));
-    }
-
-    let future = self
-        .client
-        .get(format!("{}/channel/list?{}", self.sentry_url, query.join("&")).as_str())
-        .send()
-        .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
-
-    await!(future.compat())
-}
-
-
-#[derive(Clone)]
-pub(crate) struct SentryApi {
+#[derive(Clone, Debug)]
+pub struct SentryApi<T: Adapter> {
+    pub validator: ValidatorDesc,
+    pub adapter: T,
     pub sentry_url: String,
     pub client: Client,
     pub logging: bool,
-    pub channel: Channel
+    pub channel: Channel,
+    pub config: Config,
 }
 
-impl SentryApi {
-
-    pub new(adapter: impl Adapter, channel: &Channel, logging: bool) -> Self {
+impl<T: Adapter + 'static> SentryApi<T> {
+    pub fn new(
+        adapter: T,
+        channel: &Channel,
+        config: &Config,
+        logging: bool,
+    ) -> Result<Self, ValidatorWorkerError> {
         let whoami = adapter.whoami();
-        let validator = channel.spec.validators.find(|v| v.id === whoami);
-        // assert!(vaildator, 'we can not find validator entry for whoami')
-        let sentry_url = format!("{}/channel/{}", validator.url, channel.id);
-
-        Self {
-            sentry_url,
-            client: Client::new(),
-            logging,
-            channel: channel.to_owned()
+        let validator = channel
+            .spec
+            .validators
+            .into_iter()
+            .find(|&v| v.id == whoami);
+
+        let client = Client::builder()
+            .timeout(Duration::from_secs(config.fetch_timeout.into()))
+            .build()
+            .unwrap();
+
+        match validator {
+            Some(v) => {
+                let sentry_url = format!("{}/channel/{}", v.url, channel.id);
+
+                Ok(Self {
+                    validator: v.clone().to_owned(),
+                    adapter,
+                    sentry_url,
+                    client,
+                    logging,
+                    channel: channel.to_owned(),
+                    config: config.to_owned(),
+                })
+            }
+            None => Err(ValidatorWorkerError::InvalidValidatorEntry(
+                "we can not find validator entry for whoami".to_string(),
+            )),
         }
     }
 
-    pub fn propagate() {
+    pub fn propagate(&self, messages: Vec<MessageTypes>) {
+        let serialised_messages: Vec<String> = messages
```

Comment:
why serialize the message here and pass a String?

Reply:
cos it's much easier and you don't have to serialized messages in both the leader and follower

Reply:
But you are passing the serialized message to `propagate_to` which is not public? This has nothing to do with the `leader`/`follower`, I just meant that you can use the MessageType and encode it when you are sending the request, not before that and loosing all the type-safety stuff

Reply:
There is no losing type safety stuff. The propagate function is public and it has the appropriate type definition. The propagate_to function does one thing only takes in an array of encoded messages and posts it a validators url